controllers/version_update.go

```diff
@@ -0,0 +1,489 @@
+package controllers
+
+import (
+	v1alpha1 "cdap.io/cdap-operator/api/v1alpha1"
+	"fmt"
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"log"
+	"reflect"
+	"sigs.k8s.io/controller-reconciler/pkg/reconciler"
+	"sigs.k8s.io/controller-reconciler/pkg/reconciler/manager/k8s"
+	"sigs.k8s.io/controller-reconciler/pkg/status"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type VersionUpdated = bool
+
+var (
+	updateStatus *VersionUpdateStatus
+)
+
+func init() {
+	updateStatus = new(VersionUpdateStatus)
+	updateStatus.init()
+}
+
+/////////////////////////////////////////////////////////////
+///// Main functions for handle image upgarde/downgrade /////
+/////////////////////////////////////////////////////////////
+
+func handleVersionUpdate(master *v1alpha1.CDAPMaster, labels map[string]string, observed []reconciler.Object) ([]reconciler.Object, error) {
+	// Let the current update complete if there is any
+	if isConditionTrue(master, updateStatus.Inprogress) {
+		log.Printf("Version update ingress. Continue... ")
+		return upgradeForBackend(master, labels, observed)
+	}
+
+	if objs, versionUpdated, err := updateForUserInterface(master); err != nil {
+		return nil, err
+	} else if versionUpdated {
+		return objs, nil
+	}
+
+	// Update backend service image version
+	curVersion, err := getCurrentImageVersion(master)
+	if err != nil {
+		return nil, err
+	}
+	newVersion, err := getNewImageVersion(master)
+	if err != nil {
+		return nil, err
+	}
+	if len(curVersion.rawString) == 0 {
+		setImageToUse(master)
+		return []reconciler.Object{}, nil
+	}
+
+	switch compareVersion(curVersion, newVersion) {
+	case -1:
+		// Upgrade:
+		updateStatus.clearAllConditions(master)
```

Comment:
Is this correct? Won't this clear the status each reconcile loop, thus defeating the purpose of tracking the state.

Reply:
Updated the code based on offline discussion. 

At the high level, make cdap operator not retrying upgrade. So if upgrade fails, leave all conditions in their state for user to exam. 

So the logic to check the upgrade status and react to it is:
if (upgrade inflight) { wait }
if (upgrade failed condition is set) {
  if (spec.image == status.image) {
     return upgrade_failed
  } else {
    return bad_state (because image upgraded but post-upgrade processing failed)
  }
}
return upgrade_succeeded.

Reply:
lgtm